external/fv3fit/fv3fit/pytorch/graph.py

```diff
@@ -0,0 +1,183 @@
+        output_variables: names of variables to use as outputs
+        optimizer_config: selection of algorithm to be used in gradient descent
+        graph_network: configuration of graph network
+        training_loop: configuration of training loop
+        loss: configuration of loss functions, will be applied separately to
+            each output variable
+    """
+
+    input_variables: List[str]
+    output_variables: List[str]
+    normalization_fit_samples: int = 50_000
+    optimizer_config: OptimizerConfig = dataclasses.field(
+        default_factory=lambda: OptimizerConfig("AdamW")
+    )
+    build_graph: GraphConfig = dataclasses.field(default_factory=lambda: GraphConfig())
+
+    graph_network: GraphNetworkConfig = dataclasses.field(
+        default_factory=lambda: GraphNetworkConfig()
+    )
+    training_loop: TrainingLoopConfig = dataclasses.field(
+        default_factory=lambda: TrainingLoopConfig()
+    )
+    loss: LossConfig = LossConfig(loss_type="mse")
+
+    @property
+    def variables(self) -> Set[str]:
+        return set(self.input_variables).union(self.output_variables)
+
+
+def get_normalizer(sample: Mapping[str, np.ndarray]):
+    scalers = {}
+    for name, array in sample.items():
+        s = StandardScaler()
+        s.fit(array)
+        scalers[name] = s
+
+    def scale(sample: Mapping[str, np.ndarray]):
+        output = {**sample}
+        for name, array in sample.items():
+            output[name] = scalers[name].normalize(array)
+        return output
+
+    return scale
+
+
+# TODO: Still have to handle forcing
+
+
+@register_training_function("graph", GraphHyperparameters)
+def train_graph_model(
+    hyperparameters: GraphHyperparameters,
+    train_batches: tf.data.Dataset,
+    validation_batches: Optional[tf.data.Dataset],
+):
+
+    """
+    Train a graph network.
+
+    Args:
+        train_batches: training data, as a dataset of Mapping[str, tf.Tensor]
+        validation_batches: validation data, as a dataset of Mapping[str, tf.Tensor]
+        build_model: the function which produces the pytorch model
+            from input and output samples. The models returned must take a list of
+            tensors as input and return a list of tensors as output.
+        input_variables: names of inputs for the pytorch model
+        output_variables: names of outputs for the pytorch model
+        n_epoch: number of epochs
+    """
+
+    # use transforms to get correct variables, correct dimensions
+    processed_train_batches = train_batches.map(apply_to_mapping(ensure_nd(3)))
+
+    sample = next(
+        iter(
+            processed_train_batches.unbatch().batch(
+                hyperparameters.normalization_fit_samples
+            )
+        )
+    )
+
+    normalizer = get_normalizer(sample)
+
+    get_Xy = curry(get_Xy_dataset)(
+        input_variables=hyperparameters.input_variables,
+        output_variables=hyperparameters.output_variables,
+        n_dims=3,
```

Comment:
I'm a little confused that `n_dims` is `3` - don't you need `tile, x, y, z` so 4 dimensions? The input data will have these dimensions, though scalars will be missing the z-dimension.